### PR TITLE
Fixes problem with notices cache

### LINKED ISSUE #14072: `conda notices` doesn't display notice when package not found and considers it as seen latest
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

notices are pulled from the server but not displaying because of PackageNotFoundError
after correcting package name notices are not displayed because they were "seen" already

1. Checking cache folder
```
ls -al <HOME>/Library/Caches/conda/notices
total 0
drwxr-xr-x  3 user group  96 Jul 24 12:46 .
drwxr-xr-x  3 user group  96 Oct 14  2022 ..
-rw-r--r--  1 user group   0 Jul 22 00:00 notices.cache
```

2. Installing unknown package `absl-py2`:
```
conda install absl-py2
Retrieving notices: ...working... done
Channels:
 - http://localhost:8000
 - defaults
Platform: osx-64
Collecting package metadata (repodata.json): done
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - absl-py2

Current channels:

  - http://localhost:8000
  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.
```

3. Checking cache folder again:
```
ls -al <HOME>/Library/Caches/conda/notices
total 8
drwxr-xr-x  4 user group  128 Jul 24 12:46 .
drwxr-xr-x  3 user group   96 Oct 14  2022 ..
-rw-r--r--  1 user group  207 Jul 24 12:46 306295661706d0ba4a0aaa5f95e0dcfb0378d9fdb7914bd17b0f1f00372117f5.json
-rw-r--r--  1 user group    0 Jul 24 12:46 notices.cache
```

4. Installing correct package and don't see notices
```
conda install absl-py
Channels:
 - http://localhost:8000
 - defaults
Platform: osx-64
Collecting package metadata (repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /opt/miniconda3

  added / updated specs:
    - absl-py


The following NEW packages will be INSTALLED:

  absl-py            /noarch::absl-py-0.11.0-pyhd3eb1b0_1


Proceed ([y]/n)? y


Downloading and Extracting Packages:

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
```

5. Can see notices if run `conda notices` 

### Conda Info

```shell
active environment : <REDACTED>
    active env location : <REDACTED>
            shell level : 2
       user config file : <REDACTED_HOME>/.condarc
 populated config files : <REDACTED_HOME>/.condarc
          conda version : 24.5.0
    conda-build version : 24.5.1
         python version : 3.9.18.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=m1
                          __conda=24.5.0=0
                          __osx=10.16=0
                          __unix=0=0
       base environment : /opt/miniconda3  (writable)
      conda av data dir : /opt/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : http://localhost:8000/osx-64
                          http://localhost:8000/noarch
          package cache : /opt/miniconda3/pkgs
                          <REDACTED_HOME>/.conda/pkgs
       envs directories : /opt/miniconda3/envs
                          <REDACTED_HOME>/.conda/envs
               platform : osx-64
             user-agent : conda/24.5.0 requests/2.31.0 CPython/3.9.18 Darwin/22.6.0 OSX/10.16 solver/libmamba conda-libmamba-solver/24.1.0 libmambapy/1.5.6 aau/0.4.3
                UID:GID : 502:20
             netrc file : None
           offline mode : False
```

### Conda Config

```shell
channels:
  - http://localhost:8000

==> envvars <==
allow_softlinks: False
```

### Conda list

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by travishathaway**
### Description

Fixes: #14072 

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: conda/notices/cache.py [hunk 1]
```diff
@@ -148,3 +164,15 @@ def get_viewed_channel_notice_ids(
...
+def clear_cache() -> None:
+    """
+    Removes all files in notices cache
+    """
+    cache_dir = get_notices_cache_dir()
+
+    for cache_file in os.listdir(cache_dir):
+        full_cache_file_path = Path(cache_dir) / cache_file
+        if full_cache_file_path.is_file():
+            full_cache_file_path.unlink()
```
Comment (kenodegard) [inline conda/notices/cache.py:178]:
```suggestion
    cache_dir = Path(get_notices_cache_dir())

    for cache_file in cache_dir.iterdir():
        if cache_file.is_file():
            cache_file.unlink()
```

File: conda/notices/core.py [hunk 2]
```diff
@@ -124,10 +124,19 @@ def wrapper(*args, **kwargs):
...
+                try:
+                    return_value = func(*args, **kwargs)
+                    display_notices(channel_notice_set)
+
+                    return return_value
+
+                except Exception:
+                    try:
+                        # Remove the notices cache file if we encounter an exception
+                        cache.clear_cache()
```
Comment (travishathaway) [inline conda/notices/core.py:136]: I'm still not sure this is the right approach to use. This means that every time an error is encountered, the cache files get removed and then conda checks again for these notices which is probably not the desired behavior 😭.

File: conda/notices/cache.py [hunk 3]
```diff
@@ -81,14 +87,24 @@ def get_notices_cache_dir() -> Path:
 
 
 def get_notices_cache_file() -> Path:
-    """Returns the location of the notices cache file as a Path object"""
+    """
+    Returns the location of the notices cache file as a Path object
+
+    If the file does not exist, we create it and set the "created_at" time stamp
+    of the file so that it is in the past. This forces the notices to be checked
+    for and read later.
```
Comment (kenodegard) [inline conda/notices/cache.py:95]: This feels like a comment more than docs. I'd shorten this to:

```suggestion
    Returns the location of the notices cache file as a Path object,
    creating it if the files does not exist.
```

I'd then put the technical comment below:

```python
# set the `created_at` timestamp in the past to force checking the notices
current_time = time.time()
...
```

File: conda/notices/core.py [hunk 4]
```diff
@@ -124,10 +124,19 @@ def wrapper(*args, **kwargs):
...
+                    except OSError:
```
Comment (kenodegard) [inline conda/notices/core.py:137]: Hm maybe `clear_cache` should handle `OSError` for us?

File: tests/cli/test_main_notices.py [hunk 5]
```diff
@@ -409,11 +380,7 @@ def test_notices_does_not_interrupt_command_on_failure(
     prefix = path_factory()
 
     _, _, exit_code = conda_cli(
-        "create",
-        f"--prefix={prefix}",
-        "--yes",
-        *("--channel", "local"),
-        "--override-channels",
+        "create", f"--prefix={prefix}", "--yes", "--channel", test_recipes_channel
```
Comment (kenodegard) [inline tests/cli/test_main_notices.py:383]:
```suggestion
        "create",
        f"--prefix={prefix}",
        "--yes",
        f"--channel={test_recipes_channel}",
```

File: tests/cli/test_main_notices.py [hunk 6]
```diff
@@ -440,4 +407,52 @@ def test_notices_cannot_read_cache_files(
...
+    messages_json = get_test_notices(messages)
+    add_resp_to_mock(notices_mock_fetch_get_session, 200, messages_json)
+
+    with pytest.raises(PackagesNotFoundError):
+        conda_cli(
+            "create",
+            "--name",
+            env_one,
+            "--channel",
+            test_recipes_channel,
```
Comment (kenodegard) [inline tests/cli/test_main_notices.py:438]:
```suggestion
            f"--name={env_one}",
            f"--channel={test_recipes_channel}",
```

File: tests/cli/test_main_notices.py [hunk 7]
```diff
@@ -440,4 +407,52 @@ def test_notices_cannot_read_cache_files(
...
+    messages = ("Test One",)
+    messages_json = get_test_notices(messages)
+    add_resp_to_mock(notices_mock_fetch_get_session, 200, messages_json)
+
+    out, err, exc = conda_cli(
+        "create",
+        "--name",
+        env_one,
+        "--channel",
+        test_recipes_channel,
```
Comment (kenodegard) [inline tests/cli/test_main_notices.py:453]:
```suggestion
        f"--name={env_one}",
        f"--channel={test_recipes_channel}",
```

**Review by kenodegard — DISMISSED** [review]

File: conda/notices/core.py (hunk 4 above)
Comment (travishathaway) [reply conda/notices/core.py:137]: I would say no because it should be up to the calling code to decide whether the exception can be ignored or not.

File: conda/notices/core.py (hunk 2 above)
Comment (travishathaway) [reply conda/notices/core.py:136]: Any type of error because if any error occurs at all during the command that is decorated with this decorator, we don't want the notices to be shown. At that point, there's so much output on the screen that it would be too much information presented.

File: conda/notices/cache.py [hunk 8]
```diff
@@ -81,14 +87,24 @@ def get_notices_cache_dir() -> Path:
 
 
 def get_notices_cache_file() -> Path:
-    """Returns the location of the notices cache file as a Path object"""
+    """
+    Returns the location of the notices cache file as a Path object
```
Comment (dholth) [inline conda/notices/cache.py:91]:
```suggestion
    Return path of notices cache
```
https://peps.python.org/pep-0257/#one-line-docstrings

Type is given as annotation, doesn't need to be repeated in docstring.

File: conda/notices/cache.py [hunk 9]
```diff
@@ -81,14 +87,24 @@ def get_notices_cache_dir() -> Path:
...
     cache_dir = get_notices_cache_dir()
     cache_file = cache_dir.joinpath(NOTICES_CACHE_FN)
 
     if not cache_file.is_file():
         with open(cache_file, "w") as fp:
             fp.write("")
 
+        current_time = time.time()
+        create_time = current_time - NOTICES_DECORATOR_DISPLAY_INTERVAL
+        os.utime(cache_file, (create_time, create_time))
```
Comment (dholth) [inline conda/notices/cache.py:106]: This seems really odd. Can we write the timestamp to a file that has true filesystem timestamps (i.e. the current time?) Recommend using mtime only as the timestamp conda cares about.

Comment (jezdez) [reply conda/notices/cache.py]: I've done this in [`f0d9245` (#14229)](https://github.com/conda/conda/pull/14229/commits/f0d9245c173509ff635cbfcc031f60d2db4e3079)

File: news/14229-bugfix-for-notices-cache [hunk 10]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fixes bug with notices cache (#14229)
```
Comment (jezdez) [inline news/14229-bugfix-for-notices-cache]:
```suggestion
* Improve conda notices cache handling for better user experience. (#14229)
```

**Review by travishathaway — COMMENTED** [review]
I'm feeling good about the changes that @jezdez has made since I started working on this. I think it's ready to be merged.

**Review by jaimergp — DISMISSED** [review]

File: tests/notices/test_core.py [hunk 11]
```diff
@@ -137,8 +137,10 @@ def test__conda_user_story__disable_notices(
     I do not want to see notifications while running commands like,  "install",
     "update" or "create".
     """
+    # The disable_channel_notices fixture has already set number_channel_notices to 0
+    # We just need to set the channels without resetting the context
     monkeypatch.setenv("CONDA_CHANNELS", "defaults")
-    reset_context()
+    # Don't call reset_context() here as it would override the fixture's settings
```
Comment (kenodegard) [inline tests/notices/test_core.py]: IMO this means `disable_channel_notices` should be refactored

**Review by kenodegard — DISMISSED** [review]

File: tests/notices/test_core.py (hunk 11 above)
Comment (jezdez) [reply tests/notices/test_core.py]: Removed it in [`91a24cb` (#14229)](https://github.com/conda/conda/pull/14229/commits/91a24cb33622645de91e7fe5c72d502f81d3e299)

**Review by jezdez — APPROVED** [review]
